modules/hagrid.nix

```diff
@@ -0,0 +1,154 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (builtins) readFile toString;
+  inherit (lib) attrsets generators modules options strings types;
+  inherit (attrsets) mapAttrsToList recursiveUpdate;
+  inherit (generators) toINI;
+  inherit (modules) mkIf;
+  inherit (options) literalExpression mdDoc mkEnableOption mkOption;
+  inherit (strings) concatStrings replaceStrings;
+
+  cfg = config.services.hagrid;
+  toml = pkgs.formats.toml {};
+  defaultSettings = {
+    global = {
+      address = "0.0.0.0";
+      port = 8080;
+    };
+    release = {
+      assets_dir = "${cfg.webRoot}/assets";
+      base-URI = "https://${cfg.hostname}";
+      base-URI-Onion = "https://${cfg.hostname}";
+      email_template_dir = "${cfg.dataDir}/email-templates";
+      enable_prometheus = false;
+      from = "noreply@localhost";
+      keys_external_dir = "${cfg.webRoot}/keys";
+      keys_internal_dir = "${cfg.dataDir}/keys";
+      mail_rate_limit = 3600;
+      maintenance_file = "${cfg.dataDir}/maintenance";
+      template_dir = "${cfg.dataDir}/templates";
+      tmp_dir = "${cfg.dataDir}/tmp";
+      token_dir = "${cfg.dataDir}/tokens";
+      token_secret = "$TOKEN_SECRET";
+      token_validity = 3600;
+      x-accel-redirect = true;
+    };
+  };
+in {
+  meta.maintainers = with lib.maintainers; [ ];
+
+  options = {
+
+    services.hagrid = {
+
+      enable = mkEnableOption (mdDoc ''
+        hagrid (Verifying OpenPGP keyserver, written in Rust)'');
+
+      package = mkOption {
+        default = pkgs.hagrid;
+        defaultText = literalExpression "pkgs.hagrid";
+        type = types.package;
+        description = mdDoc "Which hagrid derivation to use.";
+      };
+
+      dataDir = mkOption {
+        type = types.path;
+        default = "/var/lib/hagrid";
+        example = "/var/db/hagrid";
+        description = mdDoc ''
+          Data directory for hagrid, where the keys, assets,
+          templates and Rocket.toml are located.
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = types.path;
+        description = mdDoc ''
+          Systemd EnvironmentFile with secrets to inject
+          into the service config before startup.
+        '';
+      };
+
+      hostname = mkOption {
+        type = types.str;
+        description = mdDoc ''
+          Which hostname to set the base-URI of hagrid to.
+        '';
+      };
+
+      settings = mkOption {
+        type = toml.type;
+        default = defaultSettings;
+        apply = settings: recursiveUpdate defaultSettings settings;
+        description = mdDoc ''
+          Configuration for hagrid rocket server in toml type (attrset).
+          See https://gitlab.com/keys.openpgp.org/hagrid/ for details.
+        '';
+      };
+
+      webRoot = mkOption {
+        type = types.path;
```

Comment:
Using `types.path` is IMHO rather dangerous for that. This also allows native Nix paths to be passed to this option that will turned into store-paths during evaluation. I'd suggest to use `types.str` instead. Same applies to `dataDir`.